### DISCUSSION TITLE
[Closed] Xmlns Definition

### DISCUSSION THREAD

**Discussion by dansiegel** [CommunityToolkit/Maui#144] [New Feature Discussions]
As previously discussed for the Xamarin.CommunityToolkit it is highly useful when building Views in XAML to have a singular XmlnsDefinition that you can import. The same sort of thing appears to be currently missing and would be highly useful for the Maui Community Toolkit.

```xml
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:toolkit="http://schemas.microsoft.com/dotnet/2021/maui/tookit"
             x:Class="MauiApp2.MainPage">
</ContentPage>
```

If this sounds good happy to submit the PR for it

**Comment by TheCodeTraveler** [comment]
Agreed! I love it! 

Just FYI - before opening a PR, the [next step is to open a Proposal](https://github.com/CommunityToolkit/Maui#submitting-a-new-feature).

@jsuarezruiz + @jfversluis - as .NET MAUI devs/experts, are there any downsides, gotchyas or best practices for implementing a singular XmlnsDefinition?

**Reply by dansiegel** [reply]
what we found around the time I added it to the Xamarin.CommunityToolkit was that we needed to have a dummy attribute that got brought in silently to keep the linker from linking out the Toolkit if your only references were in XAML... if the Toolkit has an extension to the builder like:

```cs
builder.UseCommunityToolkit()
```

Then that hack wouldn't be needed for the Maui implementation.

**Reply by jfversluis** [reply]
Is that what prevents this? https://github.com/xamarin/XamarinCommunityToolkit/issues/1311

Then I’m all for it. But if we’re going to have this all over I would rather find a solution for this first.

**Reply by dansiegel** [reply]
I believe the linker would cause a similar issue.. I have also seen the Xamarin tooling just start failing with vanilla Xamarin.Forms and give that error though as well... If it helps for context, this was our work around in XCT for the linker issue:
https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/build/Xamarin.CommunityToolkit.targets

and again having an extension on the Builder negates the need for the same solution we used in XCT

**Reply by TheCodeTraveler** [reply]
Just FYI - we discussed this in our monthly standup today: https://www.youtube.com/watch?v=Oxm_io0k1dY&fbclid=IwAR1kCMRjnmZU3eYlc60yy829WJyaPl4yjgvUTog9WvywejUHyqm6OGuzh3I

@pictos Is going to look into using Source Generators to avoid the error that Gerald is referring to.

I think we're all on board for adding this. We'd prefer to avoid an initialization step, e.g. `builder.UseCommunityToolkit()`, but we can do it if that's the only path forward.

**Reply by dansiegel** [reply]
how exactly are you registering the Handlers for the Toolkit? As I recall from my custom Icon Image Source I had to integrate with the builder to do the registrations. Regardless of whether someone manually calls UseCommunityToolkit() or there is some code gen, that should be enough to create enough of a link to prevent the assembly from getting stripped out at build.

**Reply by jfversluis** [reply]
Aah that's right. We don't use reflection anymore for the renderers and we have to register the handlers now anyway so a `UseCommunityToolkit()` call would be warranted as it actually does something, right?

**Reply by dansiegel** [reply]
correct with Maui it would actually be doing something useful

**Reply by pictos** [reply]
Just to clarify, my English could lead in some confusion about my idea. The proposal for using the Source Generator (SG) is to fix the issue in XCT, using the compiler on my head, I believe that will be enough to make VS happy and doesn't pop that fake error.

We can use SG for the MCT as well, the Maui team was spiking about using SG to avoid the boilerplate code in some areas. If there's some point of extension for us we can use it to inject our `builder.UseCommunityToolkit()`, if the user doesn't need to do anything fancy.

-  https://github.com/dotnet/maui/pull/2501

**Reply by dansiegel** [reply]
Ultimately it shouldn't much matter whether the code is from a Source Generator or the user has to manually do it. I think the bigger take away here is that Maui relies on registrations with the Builder vs runtime Reflection, so regardless there will be code somewhere that should be enough to resolve the problem with VS. 😉

**Reply by pictos** [reply]
@dansiegel yes, but our goal is to avoid the user typing the code if we can. For XCT will be great if I can make it work

**Reply by jfversluis** [reply]
Circling back to this. One scenario that isn't covered, or rather, multiple scenarios, is when people just use a converter or something that doesn't need the `UseMauiToolkit` call to register handlers.

Anyway, @dansiegel are you willing and able to write this up with some considerations we talked about and from your pov in a more formal proposal? Not to rush you, but if you can get it done by Dec. 2nd we can talk about it on the live stream for that day as well :)

Thanks! <3

**Comment by bijington** [comment]
This is now complete as part of #247.

No clue on how to close this though

**Reply by TheCodeTraveler** [reply]
There isn't really a way to close these. I recommend just editing the title and prepending `[Closed]`.

Here's an example of one that I "closed": https://github.com/CommunityToolkit/Maui/discussions/33

**Reply by bijington** [reply]
Ha I saw that one the other day and thought it was closed. This will do then :)

**Comment by CliffAgius** [comment]
Closed as answered.
